docs/features/non-source-outputs.md

```diff
@@ -0,0 +1,141 @@
+- The user can control where the generated artifacts are emitted with a suitable default
+- Be optimal in terms of file writes / copies and ensure we keep the correct incremental build semantics for MSBuild
+- Do not have to be run as part of the IDE experience. Files are only emitted to disk as part of command line compilation
+
+## Implementation
+
+The feature is exposed to incremental generator authors via a new output API:
+
+```csharp
+namespace Microsoft.CodeAnalysis;
+
+public readonly partial struct IncrementalGeneratorInitializationContext
+{
+    public void RegisterArtifactOutput<TSource>(IncrementalValueProvider<TSource> source, Action<ArtifactProductionContext, TSource> action);
+    public void RegisterArtifactOutput<TSource>(IncrementalValuesProvider<TSource> source, Action<ArtifactProductionContext, TSource> action);
+}
+
+public readonly struct ArtifactProductionContext
+{
+    public CancellationToken CancellationToken { get; }
+
+    public void AddFile(string hintName, Action<System.IO.Stream, CancellationToken> callback);
+}
+```
+
+This can be used by a generator author like so:
+
+```csharp
+internal sealed class Generator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var compilationSource = context.CompilationProvider;
+
+        context.RegisterArtifactOutput(compilationSource, (nspc, item) =>
+        {
+            nspc.AddFile("interop.js", (stream, cancellationToken) =>
+            {
+                using var writer = new StreamWriter(stream);
+                writer.WriteLine("//output");
+            });
+        });
+    }
+}
+```
+
+Note how the call to `AddFile` takes a callback that provides a stream. This allows the host of the generator driver to control how the stream is created and what kind of stream is created. For instance in the compiler it could be a `FileStream` that writes directly to disk, but in unit tests it could be a `MemoryStream` that can be inspected as part of the test. It also takes a `hintName`, the same way that source files do, rather than an explicit file name or path. This is explained in more detail below.
+
+In order to control whether this output is used or not, a new entry is added to the `IncrementalGeneratorOutputKind` enum:
+
+```diff
+[Flags]
+public enum IncrementalGeneratorOutputKind
+{
+    None = 0,
+    Source = 0b1,
+    PostInit = 0b10,
++    Artifact = 0b100
+}
+```
+
+A host such as the IDE may disable non-source outputs when constructing the Generator Driver.
+
+In terms of the compilation and inputs that an Artifact can be based on, artifact outputs are treated the same as source outputs; all generation is based on the state of the compilation *before* any other generation has happened, and an artifact can't use the output of another generator (outside of itself) to generate artifacts.
```

Comment:
this, to me, seems incredibly odd.

1. we have already said that this is not c# code, so it doesn't influence the compiler.
2. the use cases we're presented (like emitting a js interop api) would need to be generating against the *final* compilation.  So this seems like it's kneecapped unnecessarily from the start.

I would think we could be *very* different here.  These run on the final compilation, producing *final* artifacts for tools after the csc step.

Reply:
If we go that route then I think we have to create a 'new' kind of generator, that only produces artifacts, not add it to incremental generators. Allowing artifacts to see the output compilation would require us to re-run parts of the graph in a second pass.

Its technically feasible; we would just run fork the driver and run it a second time only for these outputs, it's just extremely confusing to explain to authors that some code runs at one time, some runs at another, and any shared parts are run twice with different inputs. 

Given that incremental generators already confuse authors, making it more obscure seems bad. If we really care about running after everything is generated I think we should make these a new thing.

Reply:
> it's just extremely confusing to explain to authors that some code runs at one time

To me, i think it's reasonable to explain it in the following fashion:

We start with compiler inputs.
We run the generator steps to add post init files.
We get the initial compilation.
We run the normal generators to produce *c# code*.
We get the final compilation.
We run analyzers and artifact-generation on the final-compilation. 

The only one that is special is the middle one, since it needs to do some amount of analysis that feeds back into the project.

--

I guess put another way: i don't see how this could work (say for the js cases we're using as examples) if they don't see generated code.  if you're trying to spit out a js-api... what do you do for things like the APIs generated by things like the JsonGenerator, etc.?

Reply:
Hmm, that's true that we do it for post init already so there is precedence.

I guess I am coming at it from the point of view that you already have a source generator, it makes some C# stuff; it also wants to make some non-C# stuff to go along with it. As opposed to, I have a new generator that wants to make non-C# stuff.

FWIW The asp.net scenarios (and I also discovered today, Gladstone) do work in this model. There are possibly generators that don't, but I don't have concrete examples yet.

Reply:
+1 for artifact generation on final compilation. When running in the IDE, would we even run this step? I would think we can skip it entirely.